# Python_Bits
1. Upset_Plot_Orthofinder.py - Python code to create upset plots from Orthogroups.GeneCount.tsv file - an output of Orthofinder tool. But this must be edited according to the number of species given by you to orthofinder
2. UpsetPlot_Orthofinder_Genecount.py - Custom python script that constructs upset plot from Orthofinder Orthogroups.GeneCount.tsv file. No edits needed. Can be used directly on the file from orthofinder.
3. Process_orthofinder_Results_correct.py, Unique_genes_annot.py, Process_orthofinder_genedup.py - python codes written for processing orthofinder results and gaining some relevant details for my study from them
4. Process_blast_results_alternaria_camptotheca.py, process_blast_campto_vs_nimmo_leaf.py, process_blast_results_alternaria_nimmoniana_leaf.py, process_blast_results_alternaria_nimmoniana_root.py, process_blast_results_alternaria_nimmoniana_stem.py, process_blast_results_alternaria_nimmoniana_lrs.py, process_blast_alternaria_candidates(1).py - python codes written for analyzing and processing results of BLASTp done with host plant and endophyte
5. Statistics_gene_prediction.py - pyhton code for obtaining some information on statistics of gene prediction performed
6. avg_prot_length.py- Script to calculate the average protein sequence length in a fasta file containing peptide sequences
